internal/registration/internal/domain/user_registration.go

```diff
@@ -0,0 +1,84 @@
+package domain
+
+import (
+	"fmt"
+	"github.com/google/uuid"
+	"net/mail"
+	"time"
+)
+
+var (
+	UserRegistrationCannotBeConfirmedMoreThanOnceError = fmt.Errorf("UserRegistration can't be confirmed more than once")
+)
+
+func RegisterNewUser(firstName, lastName, email, password string) (*UserRegistration, error) {
+	userMail, err := CreateUserEmail(email)
+	if err != nil {
+		return nil, err
+	}
+
+	return &UserRegistration{
+		ID:               UserRegistrationId{Value: uuid.New()},
+		Status:           WaitForConfirmation,
+		Email:            *userMail,
+		Name:             firstName + " " + lastName,
+		FirstName:        firstName,
+		LastName:         lastName,
+		Password:         password,
+		RegistrationDate: time.Now(),
+		ConfirmationDate: time.Time{},
+	}, nil
+}
+
+type UserRegistration struct {
+	ID               UserRegistrationId
+	Status           UserRegistrationStatus
+	Email            UserRegistrationEmail
+	Name             string
+	FirstName        string
+	LastName         string
+	Password         string
+	RegistrationDate time.Time
+	ConfirmationDate time.Time
+}
+
+func (u *UserRegistration) Confirm() error {
+	if u.Status == Confirmed {
+		return UserRegistrationCannotBeConfirmedMoreThanOnceError
+	}
+	u.Status = Confirmed
+	u.ConfirmationDate = time.Now()
+	return nil
+}
+
+type UserRegistrationEvent struct {
+	ID               UserRegistrationId
+	email            UserRegistrationEmail
+	Name             string
+	FirstName        string
+	LastName         string
+	RegistrationDate time.Time
+}
+
+type UserRegistrationId struct {
```

Comment:
In go abbreviations are always in upper case. So it's better to rename it to UserRegistrationID.